python/pdo/contract/replication.py

```diff
@@ -0,0 +1,214 @@
+                break
+
+            #report replication completion in order to add the task to the transaction submission queue
+            task['response_object'].report_replication_completion()
+    
+    # -----------------------------------------------------------------
+
+    @classmethod
+    def __update_set_of_unreliable_storage_services__(cls):
+        """ Update the set of unreliable storage services based on prior replication attempts. Any service that returned an exception of any kind in 
+        the past is considered unreliable for the rest of the execution."""
+        
+        temp = True
+
+        while temp:
+            try: 
+                url = cls.__queue_of_SSURLS_raising_exceptions__.get_nowait()
+                if url not in cls.__urls_of_storage_services_to_ignore__:
+                    cls.__urls_of_storage_services_to_ignore__.append(url)
+                    logger.info("Adding Storage Service at %s to the list of unreliable storage services to be ignored for replication", str(url))
+            except: 
+                temp = False
+    
+    # -----------------------------------------------------------------
+
+    def __init__(self, replication_params, storage_clients, \
+                contract_id, blocks_to_replicate, enclave_service, data_dir=None):
+        """ Create an instance of the Replicator class: used for replicating the current change set"""
+        
+        self.storage_clients_to_replicate = []
+
+        if len(storage_clients) == 0:
+            return
+        
+        self.sservice_url_for_contract = enclave_service.storage_service_url
+        self.num_provable_replicas = replication_params['num_provable_replicas']
+        self.availability_duration = replication_params['availability_duration']
+        self.contract_id = contract_id
+        self.blocks_to_replicate = blocks_to_replicate
+        self.data_dir = data_dir
+
+        # identify reliable storage services to be used with this instance of Replicator
+        Replicator.__update_set_of_unreliable_storage_services__()
+        for client in  storage_clients:
+            use_client = True
+            if are_the_urls_same(client.ServiceURL, self.sservice_url_for_contract):
+                use_client = False
+            else:
+                for url in Replicator.__urls_of_storage_services_to_ignore__:
+                    if are_the_urls_same(client.ServiceURL, url):
+                        use_client = False
+                        break
+            if use_client:
+                self.storage_clients_to_replicate.append(client)
+         
+        # make sure that the number of reliable storage services is sufficient for replication 
+        if len(self.storage_clients_to_replicate) < replication_params['num_provable_replicas']-1: # -1 since we do not explicitly replicate to 
+                    # the sservice associated with the eservice
```

Comment:
why are you using the dictionary "replication_params" when you already have the num_provable_replicas attribute?

Reply:
thank you, yes, i have fixed this